how_to_solve_a_problem.md

```diff
@@ -0,0 +1,90 @@
+| Your new shiny code doesn't work. |  Check if it is even invoked. Again, no need to have a very delicate debugging session. Instead, trace the places the code is being called and figure out if is called as you assume. |
+| Something that used to work stopped working. | You are in luck, because you can try and find the working instance (e.g. go back in Git's history). If you are able to find a working example, the next step is to pin point the exact time in history (i.e. the git commit) that broke things, and then you'll have a smaller diff to investigate. |
+
+The gist of the above examples is to figure out as quickly as possible the right area of the bug. From there start pin pointing the bug's cause - there is always *one specific line* that causes the havoc.
+
+In summary: After each of those above solutions we asked ourself, did it change anything?
+If yes, we know we are in the right area, and we keep zooming in.
+If not, we know we were barking up the wrong tree. We keep using our D9 to go over other parts of the code until we see a change.
+
+![Caterpillar D9](images/solving_problem/image1.jpg)
+
+### Time boxing
+
+Like any task you must be disciplined and time box your efforts. Sometimes the _root cause_ can be buried very deep inside a legacy or a complex system. For example, trying to find the root cause in Drupal's API can be very challenging. There are times it's needed, but there are cases a work around would be easier.
+
+The way to decide when to move and develop a work around solution, is with time boxing. If you are about to exceed the time allocated, it may be time to also investigate what are the work around options.
+
+In any case, as always, be sure to reach out to other team members to see if they have a similar experience or a good suggestion.
+
+To emphasize, a work around is a valid solution, but it's kept as a last resort. Having a workaround in place often adds a technical debt to the system. Meaning, you have solved a problem, but most likely created new future ones, so be careful with this solution.
+
+## Taking responsibility
+
+Digging into a problem might reveal unexpected stuff, maybe a bug lurking in the shadows.
+Don't be distracted by it, as our brain can't process two problems at once, however don't disregard it. You are expected to open an issue in GitHub, for you or others to address it later on.
+
+The term "taking responsibility" refers to owning the code or a piece of work that was done, even if it was authored by others. Nothing is carved in stone, and past developers have not necessarily been smarter than you.
+
+A system is a living thing, that changes over time. You don't need to work around other's code, you can go ahead and fix the problem. A classic example is CSS code, which becomes bloated very quickly. Don't keep adding new classes, if with a slight more effort you can change existing classes or even completely remove them, thus reducing the CSS complexity, then do it!
+
+## Escalation
```

Comment:
This whole section speaks to communicating with the client in the correct manner. What I am missing is when to escalate internally.  Maybe something like this: "Internal escalation should use the same time boxing principle to a bug work around. If you see that you are not able to resolve the issue with in the time box you set, then its time to communicate this internally." Thoughts?